lms/djangoapps/instructor/views/api.py

```diff
@@ -392,6 +398,42 @@ def register_and_enroll_students(request, course_id):  # pylint: disable=too-man
             name = student[NAME_INDEX]
             country = student[COUNTRY_INDEX][:2]
 
+            cohort = None  # Optional[CourseUserGroup]
+            course_mode = default_course_mode
+            if len(student) > COHORT_INDEX:
+                # get a cohort name if exists, with caching in `cohorts`
+                cohort_name = student[COHORT_INDEX]
+                if cohort_name and not is_course_cohorted(course_id):
+                    row_errors.append({
+                        'username': username,
+                        'email': email,
+                        'response': _(u'Course is not cohorted but cohort provided. Ignoring cohort.')
+                    })
+                    cohort = None
+                elif cohort_name and cohort_name in cohorts:
+                    cohort = cohorts[cohort_name]
+                elif cohort_name:
+                    # if cohort doesn't exist; ignore
+                    cohort = CourseUserGroup.objects.filter(
+                        course_id=course_id,
+                        group_type=CourseUserGroup.COHORT,
+                        name=cohort_name
+                    ).first()
+                    if cohort is None:
+                        row_errors.append({
+                            'username': username,
+                            'email': email,
+                            'response': _(u'Cohort name not found: {cohort}. Ignoring cohort.'.format(cohort=cohort_name))
+                        })
+
+                    cohorts[cohort_name] = cohort
+                else:
+                    cohort = None
+
+            if len(student) > MODE_INDEX:
+                course_mode = student[MODE_INDEX] if student[MODE_INDEX] else default_course_mode
+                # TODO: check that course_mode is valid (may or may not be needed?)
```

Comment:
For what it was worth, I was able to enrol students with invalid track names, but it would be better UX to catch typos and let the user know with a warning text.